### ISSUE TITLE
dockerd unable to authenticate tls connections on centos 7

### ISSUE THREAD

**Issue by znull** [docker/docker#26468]
**Output of `docker version`:**

```
Client:
Version:      1.12.1
API version:  1.24
Go version:   go1.6.3
Git commit:   23cf638
Built:
OS/Arch:      linux/amd64
Server:
Version:      1.12.1
API version:  1.24
Go version:   go1.6.3
Git commit:   23cf638
Built:
OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 1.12.1
Storage Driver: devicemapper
 Pool Name: docker-8:2-34133680-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 11.8 MB
 Data Space Total: 107.4 GB
 Data Space Available: 7.261 GB
 Metadata Space Used: 581.6 kB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.147 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107-RHEL7 (2016-06-09)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null bridge host overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-327.28.3.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 16
Total Memory: 15.14 GiB
Name: qa-compile-6.example.com
ID: IPYH:XNBR:ACV5:PI4X:524O:5GT5:QJNT:7JRR:THWL:EZLS:T5YR:AZ2K
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 16
 Goroutines: 22
 System Time: 2016-09-10T19:11:06.657566071-07:00
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

This Centos 7 box is a kvm virtual machine.

**Steps to reproduce the issue:**
1. Install docker per instructions at https://docs.docker.com/engine/installation/linux/centos/
2. run `docker pull hello-world`

**Describe the results you received:**
```
Using default tag: latest
Pulling repository docker.io/library/hello-world
Error while pulling image: Get https://index.docker.io/v1/repositories/library/hello-world/images: x509: certificate signed by unknown authority
```

**Describe the results you expected:**

successful image download

**Additional information you deem important (e.g. issue happens only occasionally):**

additional information from running dockerd as `strace -f -e trace=file dockerd -D`:
```
[pid 16439] openat(AT_FDCWD, "/etc/ssl/certs/ca-certificates.crt", O_RDONLY|O_CLOEXEC) = 16
WARN[0010] Error getting v2 registry: Get https://registry-1.docker.io/v2/: x509: certificate signed by unknown authority
ERRO[0010] Attempting next endpoint for pull after error: Get https://registry-1.docker.io/v2/: x509: certificate signed by unknown authority
DEBU[0010] Trying to pull hello-world from https://index.docker.io v1
DEBU[0010] hostDir: /etc/docker/certs.d/docker.io
[pid 16439] openat(AT_FDCWD, "/etc/docker/certs.d/docker.io", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
DEBU[0010] [registry] Calling GET https://index.docker.io/v1/repositories/library/hello-world/images
ERRO[0010] Not continuing with pull after error: Error while pulling image: Get https://index.docker.io/v1/repositories/library/hello-world/images: x509: certificate signed by un
known authority
```

So we can see:
1. dockerd finds the system trusted CA certificates at `/etc/ssl/certs/ca-certificates.crt`
2. dockerd is unable to connect to registry-1.docker.io/v2/: `https://registry-1.docker.io/v2/: x509: certificate signed by unknown authority`
3. dockerd falls back to v1, which also fails with tls errors

Is there a problem with the CA certificates? No:
```
# openssl s_client -connect registry-1.docker.io:443 -CAfile /etc/ssl/certs/ca-certificates.crt
CONNECTED(00000003)
depth=2 C = US, O = GeoTrust Inc., CN = GeoTrust Global CA
verify return:1
depth=1 C = US, O = GeoTrust Inc., CN = RapidSSL SHA256 CA - G3
verify return:1
depth=0 OU = GT98568428, OU = See www.rapidssl.com/resources/cps (c)15, OU = Domain Control Validated - RapidSSL(R), CN = *.docker.io
verify return:1
---
Certificate chain
 0 s:/OU=GT98568428/OU=See www.rapidssl.com/resources/cps (c)15/OU=Domain Control Validated - RapidSSL(R)/CN=*.docker.io
   i:/C=US/O=GeoTrust Inc./CN=RapidSSL SHA256 CA - G3
 1 s:/C=US/O=GeoTrust Inc./CN=RapidSSL SHA256 CA - G3
   i:/C=US/O=GeoTrust Inc./CN=GeoTrust Global CA
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIEpDCCA4ygAwIBAgIDAyF3MA0GCSqGSIb3DQEBCwUAMEcxCzAJBgNVBAYTAlVT
MRYwFAYDVQQKEw1HZW9UcnVzdCBJbmMuMSAwHgYDVQQDExdSYXBpZFNTTCBTSEEy
NTYgQ0EgLSBHMzAeFw0xNTAzMTkxNzM0MzJaFw0xODA0MjEwMTUxNTJaMIGPMRMw
EQYDVQQLEwpHVDk4NTY4NDI4MTEwLwYDVQQLEyhTZWUgd3d3LnJhcGlkc3NsLmNv
bS9yZXNvdXJjZXMvY3BzIChjKTE1MS8wLQYDVQQLEyZEb21haW4gQ29udHJvbCBW
YWxpZGF0ZWQgLSBSYXBpZFNTTChSKTEUMBIGA1UEAwwLKi5kb2NrZXIuaW8wggEi
MA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDRyme75dbw9AxWZ8QFCwMWrrYY
SclZ6HCiEbxSNxHgg08rEfEYi46VrL+joBwlA0t5WIIxHF198NzzdXC4YeGzruY9
7osv5lPrcdeIi+Ad+fY6K0rBBOB3xdqSPPObrINZpDmWhCQjlsnM6a1Th0oSUCjI
345b84/8PH363YO+Qmnl8BWnaTcZoPzeywM9czQsMyF2bOH+dhxja/zim6iu8W34
yBhVQeQRd1QROuHcsQAX19DKTn6TXaAwIBY3xM1Bi5Zl6tueII4dOEoibw/ImR3c
H73Pk7j1Wx+rAXeeq7LwjkUCCSlKNrHFEQ2nbr0R7FH6cck1ppgM8ud1pHr9AgMB
AAGjggFOMIIBSjAfBgNVHSMEGDAWgBTDnPP800YINLvORn+gfFvz4gjLWTBXBggr
BgEFBQcBAQRLMEkwHwYIKwYBBQUHMAGGE2h0dHA6Ly9ndi5zeW1jZC5jb20wJgYI
KwYBBQUHMAKGGmh0dHA6Ly9ndi5zeW1jYi5jb20vZ3YuY3J0MA4GA1UdDwEB/wQE
AwIFoDAdBgNVHSUEFjAUBggrBgEFBQcDAQYIKwYBBQUHAwIwIQYDVR0RBBowGIIL
Ki5kb2NrZXIuaW+CCWRvY2tlci5pbzArBgNVHR8EJDAiMCCgHqAchhpodHRwOi8v
Z3Yuc3ltY2IuY29tL2d2LmNybDAMBgNVHRMBAf8EAjAAMEEGA1UdIAQ6MDgwNgYG
Z4EMAQIBMCwwKgYIKwYBBQUHAgEWHmh0dHBzOi8vd3d3LnJhcGlkc3NsLmNvbS9s
ZWdhbDANBgkqhkiG9w0BAQsFAAOCAQEATYgwOYz/w9Qyh/YPZQDZ0BdwhkX6OCX0
Mz8pP/OO+E+1RM7ZoAGwHvIaidFqh3WeCLHjGO2IId7Ff5EZUwZhiwog0R7Y838x
OCLza/2shuvjM/FPiyXDQ6q0w4rvpwsNjmYVDdYD8bCH3b8IlO2ysjgdhRYprsdU
jg6h+zK11/tXf6S5vegrgV0F62DYx0tuTTZq/HMuXvbgY2uL1sQ5jiHlzQndV9oL
YMYqJP5MkuAKzDL5u0b8mD/EHtoPkfWOIsA5i9YrAAoWRVOJHwfFfgSY+EpXpFc4
AZUPmdZGh6q1YNavRoOL/1D5aP/VBBtofj54uMbKOK8q6vxIXSyzaw==
-----END CERTIFICATE-----
subject=/OU=GT98568428/OU=See www.rapidssl.com/resources/cps (c)15/OU=Domain Control Validated - RapidSSL(R)/CN=*.docker.io
issuer=/C=US/O=GeoTrust Inc./CN=RapidSSL SHA256 CA - G3
---
No client certificate CA names sent
Server Temp Key: ECDH, prime256v1, 256 bits
---
SSL handshake has read 2914 bytes and written 373 bytes
---
New, TLSv1/SSLv3, Cipher is ECDHE-RSA-AES128-GCM-SHA256
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES128-GCM-SHA256
    Session-ID: 11A87C8A23BFCE5101A0CFC5DEC3617F798015DE5356C150BA9FD644FBBFD18C
    Session-ID-ctx:
    Master-Key: 0914472CADCB005601A4992169B4016D3C526AD873F8A19128E6ACA6B431D878CF2DC97F1D20A20960791B719F8D1BFA
    Key-Arg   : None
    Krb5 Principal: None
    PSK identity: None
    PSK identity hint: None
    TLS session ticket lifetime hint: 300 (seconds)
    TLS session ticket:
    0000 - 2e af 3c 58 ec 61 6f 20-f5 9d 5e 93 d2 00 f1 9f   ..<X.ao ..^.....
    0010 - c3 01 1c 59 19 fe 0f 52-c2 2c 64 49 97 15 c5 4c   ...Y...R.,dI...L
    0020 - 64 3f bc 04 45 f7 a8 41-ea 95 eb 0e 0c 56 f9 ed   d?..E..A.....V..
    0030 - 84 84 7f e0 3b 29 34 1c-ae d9 f6 b2 ac da 70 0b   ....;)4.......p.
    0040 - 8b 24 06 ab 1f f8 66 ae-49 c2 af ae ec 22 f5 c3   .$....f.I...."..
    0050 - a5 09 42 3e cb 3e b8 e4-c9 77 e1 16 ab 34 5e f7   ..B>.>...w...4^.
    0060 - 66 3d 62 cf 35 ff 7a d9-85 ff 5b a7 9c 6c e2 77   f=b.5.z...[..l.w
    0070 - e2 61 ea c3 53 ed 39 b4-98 d3 64 b5 99 6c 06 74   .a..S.9...d..l.t
    0080 - b1 a0 f4 8c 3f a6 4e be-21 ad a3 49 ed b1 0a c6   ....?.N.!..I....
    0090 - f3 15 e5 2b 01 02 52 9b-66 be 53 f3 36 ab e3 d1   ...+..R.f.S.6...

    Start Time: 1473560299
    Timeout   : 300 (sec)
    Verify return code: 0 (ok)
---
DONE
```

**Comment by cpuguy83** [comment]
It's a little strange. Is there a proxy involved, perhaps?
Does `dockerd` have `HTTP_PROXY` or `HTTPS_PROXY` set?

**Comment by znull** [comment]
No, no proxies involved. It turns out this was some kind of problem with certificate management on centos. If I do:
```
# rm -f /etc/ssl/certs/ca-certificates.crt
# update-ca-trust
# systemctl restart docker
```
then everything works normally.

**Comment by znull** [comment]
Though there's still no explanation for why `openssl s_client` can make the same connection using `/etc/ssl/certs/ca-certificates.crt`.

**Comment by mlaventure** [comment]
@znull if updating the CA worked for you, do we mind if we close this now?

**Comment by znull** [comment]
Sure, I'll close it. I was just writing down the resolution in case anyone else runs into this and google brings them here.

**Comment by nokios** [comment]
I know this is old. I had been spending the last two plus hours trying to figure out why openssl s_client was working (without arguments).  Your outline of steps you took allowed me to figure out that, while the Docker documentation says that on CentOS, docker would be looking in /etc/pki/tls/certs, it WASN'T.

openssl was, but docker was not, it was using /etc/ssl/certs.  once I figured that out, I made the file in this location match the one in /etc/pki/tls/certs, and my unknown authority problem went away.
